### PR TITLE
Fix rotation of attached particlespawner [no squash pls; ketchup ok]

### LINKED ISSUE #8363: Attached particle spawners no longer rotate with object
##### Issue type
 - Bug report
 
##### Minetest version
```
5.0.0 (stable)
```

##### Summary
Until 0.4.17.1, the particle position of attached particle spawners used to rotate with the object they were attached to. #4782. I use this behavior in advtrains to make the steam engines produce steam.
However, probably since either the rotation PR or the latest particlespawner fixes, this is broken and positons are world-absolute again. (I can't test right now what exactly reverted it, I'm on mobile)
Original report here:
https://forum.minetest.net/viewtopic.php?f=11&t=14726&p=345821#p345745
##### Steps to reproduce
Install advtrains, place a steam engine and check the smoke.

### PR DESCRIPTION
**Pull request by ClobberXD**
PR created from pgimeno's commit: https://gitlab.com/pgimeno/minetest/commit/16593ff796d8a4c96a60e69d1c7e6813d8ab22f3

****

This PR fixes attached particle spawners not rotating with the parent object. See #8363 for more information.

### Screenshot

screenshot_20190809_163916

### How to test

- Install and enable `advtrains` mod.
- Launch the game.
- Place a couple of tracks, and place a steam loco on a track.
- Observe that the steam particle spawner is positioned exactly on the chimney, and is fixed to the object in the same position irrespective of object's rotation (this is broken in `master`).

****

This PR is ready for review. Tested; works. Fixes #8363.

This PR also fixes code-style in `src/client/clientobject.h` and `src/client/particles.cpp` in a separate commit for ease of reviewing. Pls no squash.

### CODE REVIEW

**Comment by SmallJoker** [comment]
Thanks for this PR. The code looks good - I will test it this weekend.
If you have some time, please give them credits by either putting their name into the commit description or even changing the author.

**Comment by ClobberXD** [comment]
@SmallJoker I intended to change the author too, but smth came up IRL, and I'm going AFK for 6-7 days. If this PR is going to be merged before I return. Would you be able to change the author? Thanks in advance.

**Comment by Andrey2470T** [comment]
Like your minetest version on the git repo calls other error, any selection boxes of entitites get invisible what is impossible to fix.

File: src/client/content_cao.h [hunk 1]
```diff
@@ -180,6 +177,12 @@ class GenericCAO : public ClientActiveObject
 		return m_matrixnode->getRelativeTransformationMatrix();
 	}
 
+	inline const core::matrix4 &getAbsolutePosRotMatrix()
```
Comment (DS-Minetest) [inline src/client/content_cao.h]: Would anything speak against this? (I'm a newbie in cv-qualified member functions.)
```suggestion
	inline const core::matrix4 &getAbsolutePosRotMatrix() const
```
([`getAbsoluteTransformation`](http://irrlicht.sourceforge.net/docu/classirr_1_1scene_1_1_i_scene_node.html#a9c46ea82efd7ae30437fb7c385809050) is const.)

File: src/client/particles.cpp [hunk 2]
```diff
@@ -38,9 +39,10 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
 v3f random_v3f(v3f min, v3f max)
 {
-	return v3f( rand()/(float)RAND_MAX*(max.X-min.X)+min.X,
-			rand()/(float)RAND_MAX*(max.Y-min.Y)+min.Y,
-			rand()/(float)RAND_MAX*(max.Z-min.Z)+min.Z);
+	return v3f(
+		rand() / (float)RAND_MAX * (max.X - min.X) + min.X,
+		rand() / (float)RAND_MAX * (max.Y - min.Y) + min.Y,
+		rand() / (float)RAND_MAX * (max.Z - min.Z) + min.Z);
```
Comment (DS-Minetest) [inline src/client/particles.cpp]: As you touch this code, I want to note that, according to `numeric.h`, "VC++ rand() sucks".
https://github.com/minetest/minetest/blob/b2065756cf9c90d493846a559587e4acec4a7851/src/util/numeric.h#L214

File: src/client/particles.cpp [hunk 3]
```diff
@@ -299,16 +301,21 @@ ParticleSpawner::ParticleSpawner(
...
-		pos.rotateXZBy(attached_yaw);
-		pos += attached_pos;
+	if (attached_absolute_pos_rot_matrix) {
+		pos *= BS;
+		attached_absolute_pos_rot_matrix->transformVect(pos);
+		pos /= BS;
+		v3s16 camera_offset = m_particlemanager->m_env->getCameraOffset();
+		pos.X += camera_offset.X;
+		pos.Y += camera_offset.Y;
+		pos.Z += camera_offset.Z;
```
Comment (DS-Minetest) [inline src/client/particles.cpp]: Why not `pos += m_particlemanager->m_env->getCameraOffset();`?

File: src/client/particles.cpp (hunk 2 above)
Comment (ClobberXD) [reply src/client/particles.cpp]: Not _quite_ related to this PR, but interesting. What do you suggest I do here?

File: src/client/particles.cpp (hunk 3 above)
Comment (ClobberXD) [reply src/client/particles.cpp]: Will do

File: src/client/content_cao.h (hunk 1 above)
Comment (ClobberXD) [reply src/client/content_cao.h]: This method simply returns `irr::scene::IDummyTransformationSceneNode::getAbsoluteTransformation` which is also a `const` method. So... why not? :D

Comment (ClobberXD) [reply src/client/content_cao.h]: Done

**Comment by ClobberXD** [comment]
Fixed a bunch of issues. Will look into why selection boxes aren't visible or "interactable" with.

This PR also fixes code-style in `src/client/clientobject.h` via a separate commit for ease of reviewing. Pls no squash.

File: src/client/particles.cpp (hunk 3 above)
Comment (ClobberXD) [reply src/client/particles.cpp]: This causes a build error, as pos is a `v3f`, while `getCameraOffset` returns a `v3s16`. Probably why pgimeno manually added the components in the first place.

File: src/client/particles.cpp (hunk 2 above)
Comment (ClobberXD) [reply src/client/particles.cpp]: I could replace `rand` with `myrand` (PRNG implementation). But I prefer to do it in a different PR, and keep this one clean - it has already got 2 code-style commits apart from the bugfix.

**Comment by SmallJoker** [comment]
049256573b6fc379ab4cd1f867f33504de6abeee
9543b849703ba2e8edc0bd4696ccfb9e7888c293
96762af3ce39ebecb967e6cf1b9b8e709a105f6d
095f26a4f8743a8f6bf01a134d22c1e20beac228

**Pull request closed without merging**